### PR TITLE
Masking proxy credentials from URL when displayed in system info

### PR DESCRIPTION
**Pull request by dani-docker**
Signed-off-by: Dani Louca <dani.louca@docker.com>

If dockerd is configured to use a proxy https://docs.docker.com/config/daemon/systemd/#httphttps-proxy and if the proxy URL is configured for basic authentication, the `SystemInfo` (ex: `docker info`) shows the full proxy URL, including the credentials.
 

**- What I did**
This PR masks out the proxy username / password while building the info struct.
To make the solution robust, I used the golang `url` package to identify the credentials, replace them with asterisks, and rebuild the URL string.
I opted not to use basic strings replacement as it will require to address a bunch of corner cases which makes the solution more complex .

**- How I did it**
- convert the proxy URL string into a URL object
- overwrite the URL object's credentials with asterisks
- convert the URL to string, (this will result of an encoded string)
- decode the encoded string , if the decoder fails, simply returns the encoded string

**- How to verify it**
- configure your docker to use the following proxy URL `http://USER:PASSWORD@proxy.example.com:80/` by following https://docs.docker.com/config/daemon/systemd/#httphttps-proxy
- run `docker info | grep Proxy`
**Expected Results:**
`Http Proxy: http://*********:*********@proxy.example.com:80/`
**Actual Result:**
`Http Proxy: http://USER:PASSWORD@proxy.example.com:80/`
**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by dani-docker** [comment]
cc: @thaJeztah

**Review by AkihiroSuda — APPROVED** [review]

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! Left some comments inline.

While I can see the reason for masking this information (especially in the output of `docker info`), I have some concerns; this change could be a breaking change for those that use this information (for whatever reason; perhaps automation, or verifying daemon configuration). This information was added a long time ago (since https://github.com/moby/moby/pull/11125), so there's definitely a chance that people use this.

Masking this information in just `docker info` / the `/info` endpoint _limits_ the exposure, but does not prevent this information from leaking in other ways (this config is obtained from their corresponding `HTTP(S)_PROXY` environment variables); also taking into account that someone with access to the API already effectively is "root" on the host.

Perhaps not an issue, but worth to double check if we anticipate this being a breaking change.

**Comment by thaJeztah** [comment]
Oh, we should update the swagger file as well, and update the description for these fields, to describe that credentials are masked when using the endpoint; https://github.com/moby/moby/blob/7bfec8cd80c5e2ce548cd7ccf63a35bf60d802d3/api/swagger.yaml#L3760-L3784

Also may warrant a mention in the API changelog https://github.com/moby/moby/blob/7bfec8cd80c5e2ce548cd7ccf63a35bf60d802d3/docs/api/version-history.md (although it's not versioned, so not sure how to best describe it there)

File: daemon/info.go [hunk 1]
```diff
@@ -245,3 +246,19 @@ func operatingSystem() string {
...
+	parsedURL, err := url.Parse(rawURL)
+	if err != nil || parsedURL.User == nil {
+		return rawURL
+	}
+	parsedURL.User = url.UserPassword("*********", "*********")
+	maskedURL := parsedURL.String()
+	decodedURL, err := url.QueryUnescape(maskedURL)
+	// return encoded maskedURL in case of an error
+	if err != nil {
+		logrus.Errorf("Failed to decode masked URL, returning encoded masked URL: %v", err)
```
[the remark this answers is not in the document]
Comment (dani-docker) [reply daemon/info.go]: good point, I will change that

File: daemon/info.go [hunk 2]
```diff
@@ -245,3 +246,19 @@ func operatingSystem() string {
...
 }
+
+func maskCredentials(rawURL string) string {
+	parsedURL, err := url.Parse(rawURL)
+	if err != nil || parsedURL.User == nil {
+		return rawURL
+	}
+	parsedURL.User = url.UserPassword("*********", "*********")
+	maskedURL := parsedURL.String()
+	decodedURL, err := url.QueryUnescape(maskedURL)
```
[the remark this answers is not in the document]
Comment (dani-docker) [reply daemon/info.go]: @thaJeztah 
Right, I would assume that it's not common to for user to enter a URL encoded string, but if this is common, I can look into adding some logic to check if a string is URL encoded before decoding.

It's not limited to `xxxxxx` or `*****`, as the encoding decoding also applies to the rest of the URL, as the `.String()` func always return an encoded string.

**Comment by dani-docker** [comment]
> Oh, we should update the swagger file as well, and update the description for these fields, to describe that credentials are masked when using the endpoint;
> 
> [moby/api/swagger.yaml](https://github.com/moby/moby/blob/7bfec8cd80c5e2ce548cd7ccf63a35bf60d802d3/api/swagger.yaml#L3760-L3784)
> 
> Lines 3760 to 3784 in [7bfec8c](/moby/moby/commit/7bfec8cd80c5e2ce548cd7ccf63a35bf60d802d3)
> 
>        HttpProxy: 
>          description: | 
>            HTTP-proxy configured for the daemon. This value is obtained from the 
>            [`HTTP_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable. 
>   
>            Containers do not automatically inherit this configuration. 
>          type: "string" 
>          example: "http://user:pass@proxy.corp.example.com:8080" 
>        HttpsProxy: 
>          description: | 
>            HTTPS-proxy configured for the daemon. This value is obtained from the 
>            [`HTTPS_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable. 
>   
>            Containers do not automatically inherit this configuration. 
>          type: "string" 
>          example: "https://user:pass@proxy.corp.example.com:4443" 
>        NoProxy: 
>          description: | 
>            Comma-separated list of domain extensions for which no proxy should be 
>            used. This value is obtained from the [`NO_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) 
>            environment variable. 
>   
>            Containers do not automatically inherit this configuration. 
>          type: "string" 
>          example: "*.local, 169.254/16" 
> Also may warrant a mention in the API changelog https://github.com/moby/moby/blob/7bfec8cd80c5e2ce548cd7ccf63a35bf60d802d3/docs/api/version-history.md (although it's not versioned, so not sure how to best describe it there)

yes, good catch, I will update the swagger

**Comment by dani-docker** [comment]
I searched the moby code base and I don't see any place that reads the `HTTPProxy` or `HTTPSProxy` straight from the info struct.
I do share your concerns that some 3rd party code might be reading this info from the API and if this is the case, then we'll break their integration; but even if they do, I think it's still not correct design that we share the credentials through the `info` endpoint.

File: daemon/info.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply daemon/info.go]: True, but the other parts would be where URL-encoding would be needed from the start (probably in the original value already), whereas the `**` was what we added, and want to have presented as-is (without encoding)

[the remark this answers is not in the document]
Comment (dani-docker) [reply daemon/info.go]: @thaJeztah 
you are correct, the `String()` does not change the end of the URL.
replacing with `xxxxx` will make things much simpler

**Comment by dani-docker** [comment]
@thaJeztah 
Changed the implementation as suggested and added couple of extra unit tests.
Note: I was under the impression that `String()` encodes the rest of the URL, not only the credentials part, but running few tests showed that it's not the case;  by using `xxxxx` instead of `*****` , we no longer need to worry about decoding/encoding the result of `String()`

Another note, while testing different URLs variations; 
I noticed that the `url` method `parse` will fail to parse out credentials that are not escaped. 
for example running `parse` on this url `http://USER@docker:pa?sword@proxy.example.com:80/hello world` , will result of no password or username set, this will cause the masking URL to look like this `http://xxxxx:xxxxx@docker:pa?sword@proxy.example.com:80/hello world`

The expectation here is that user should encode those type of strings when they configure HTTP(s) Proxy, otherwise, their dial will also fail as the core logic uses the same `parse` function

**Comment by dani-docker** [comment]
Also updated the swagger but not sure how to deal with the history file without having an API version

File: api/swagger.yaml [hunk 3]
```diff
@@ -3760,19 +3760,19 @@ definitions:
       HttpProxy:
         description: |
           HTTP-proxy configured for the daemon. This value is obtained from the
-          [`HTTP_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable.
+          [`HTTP_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable and credentials are masked.
```
[the remark this answers is not in the document]
Comment (dani-docker) [reply api/swagger.yaml]: Thanks @thaJeztah , changes are made

[the remark this answers is not in the document]
Comment (dani-docker) [reply api/swagger.yaml]: thanks! change is made

File: api/swagger.yaml [hunk 4]
```diff
@@ -3760,19 +3760,19 @@ definitions:
...
 
           Containers do not automatically inherit this configuration.
         type: "string"
-        example: "http://user:pass@proxy.corp.example.com:8080"
+        example: "http://xxxxx:xxxxx@proxy.corp.example.com:8080"
       HttpsProxy:
         description: |
           HTTPS-proxy configured for the daemon. This value is obtained from the
-          [`HTTPS_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable.
+          [`HTTPS_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable and credentials are masked.
```
[the remark this answers is not in the document]
Comment (dani-docker) [reply api/swagger.yaml]: thanks! change is made

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by thaJeztah** [comment]
It's green! merging

**Pull request merged**